<!DOCTYPE html>
<html lang="en-us">
<head>
  <title>From RxJava to Kotlin Flow: Error Handling</title>

  <link rel="icon" type="image/png" href="favicon.png">
  <link rel="apple-touch-icon" href="apple-touch-icon-144-precompressed.png"/>

  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">

  <link rel="stylesheet"
      href="//cdnjs.cloudflare.com/ajax/libs/highlight.js/11.5.0/styles/androidstudio.min.css">
  <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/11.5.0/highlight.min.js"></script>
  <script>hljs.highlightAll();</script>
  <script src="https://kit.fontawesome.com/6d5aec2882.js" crossorigin="anonymous"></script>

  <script src="https://cdn.counter.dev/script.js" data-id="2d1779da-dc31-440f-8099-940355d12f4d" data-utcoffset="2"></script>
  
  <link rel="stylesheet" href="/css/style.css" />

  <link rel="me" href="https://androiddev.social/@krossovochkin" />

  
  <meta name="theme-color" content="#1e2327">
</head>

<body>

<header>
<nav>
  <ul>
    
    <li class="menu-left"><a href="/about">About</a></li>
    
    <li class="menu-left"><a href="/">Posts</a></li>
    
    <li class="menu-left"><a href="/apps">Apps</a></li>
    
    
    <li class="menu-right"><a rel="me" href="https://t.me/krossovochkin_newsletter" target="_blank"><i class="fa-brands fa-telegram fa-l"></i></a></li>
    
    <li class="menu-right"><a rel="me" href="https://stackoverflow.com/users/1533933/krossovochkin" target="_blank"><i class="fa-brands fa-stack-overflow fa-l"></i></a></li>
    
    <li class="menu-right"><a rel="me" href="https://linkedin.com/in/vasyadrobushkov" target="_blank"><i class="fa-brands fa-linkedin fa-l"></i></a></li>
    
    <li class="menu-right"><a rel="me" href="https://facebook.com/vasya.drobushkov" target="_blank"><i class="fa-brands fa-facebook fa-l"></i></a></li>
    
    <li class="menu-right"><a rel="me" href="/index.xml" target="_blank"><i class="fa-solid fa-rss fa-l"></i></a></li>
    
    <li class="menu-right"><a rel="me" href="https://androiddev.social/@krossovochkin" target="_blank"><i class="fa-brands fa-mastodon fa-l"></i></a></li>
    
    <li class="menu-right"><a rel="me" href="https://github.com/krossovochkin" target="_blank"><i class="fa-brands fa-github fa-l"></i></a></li>
    
  </ul>
</nav>

</header>

<br/>


<div class="meta">

  
    <h1><span class="title">From RxJava to Kotlin Flow: Error Handling</span></h1>
  
  
    <h3>February 24, 2020</h3>
  

</div>

<main>
<p><img src="https://images.unsplash.com/photo-1580357991342-89ec1672c98a?ixlib=rb-1.2.1&amp;auto=format&amp;fit=crop&amp;w=1357&amp;q=80" alt="Source"><em><a href="https://unsplash.com/photos/aYNyC6fIH84">Source</a></em></p>
<blockquote>
<p>This post is part of series of comparing RxJava to Kotlin Flow. Previous articles were about <a href="https://proandroiddev.com/from-rxjava-2-to-kotlin-flow-threading-8618867e1955">Threading </a>and <a href="https://proandroiddev.com/from-rxjava-to-kotlin-flow-backpressure-d1fb91e6dea8">Backpressure</a>.</p>
</blockquote>
<p><a href="https://proandroiddev.com/from-rxjava-to-kotlin-flow-error-handling-da1f6a4f2708"><img src="https://img.shields.io/badge/original-proandroiddev-green#badge" alt=""></a></p>
<h2 id="introduction">Introduction</h2>
<p>Error handling is fundamental in reactive programming. Reactive streams might fail with exception and propagate it as an event downstream notifying consumers on the error that happened.
As in previous articles, we’ll try to compare RxJava and Kotlin Flow error handling mechanisms. This might help us to migrate existing code from RxJava to Kotlin Flow and also understand how to properly use Kotlin Flow.</p>
<h2 id="rxjava-error-handling">RxJava Error Handling</h2>
<p>In RxJava general concepts for error handling are pretty simple and straightforward:</p>
<ul>
<li>
<p>“Error” is an event which might happen in a stream</p>
</li>
<li>
<p>That “Error event” is propagated downstream as a terminal event. That means that after error happened stream is basically finished and no more events can come through it.</p>
</li>
<li>
<p>Consumers are required to handle errors in the onError callback of Observer</p>
</li>
<li>
<p>If Consumer didn’t handle error in Observer callback, then that error is sent to a global error handler (which in case of Android crashes the app by default).
<strong>NOTE</strong>: some errors which happen inside stream can go directly to global error handler e.g. in cases when the stream is already disposed.</p>
</li>
<li>
<p>Not a single error can escape chain. All the exceptions will go to the onError callback of Observer (or global error handler).</p>
</li>
</ul>
<p>A basic example of error handling is:</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-kotlin" data-lang="kotlin"><span class="line"><span class="cl"><span class="n">observeChanges</span><span class="p">()</span>
</span></span><span class="line"><span class="cl">    <span class="p">.</span><span class="n">subscribe</span><span class="p">(</span>
</span></span><span class="line"><span class="cl">        <span class="p">{</span> <span class="k">value</span> <span class="o">-&gt;</span> <span class="n">println</span><span class="p">(</span><span class="s2">&#34;value: </span><span class="si">$value</span><span class="s2">&#34;</span><span class="p">)</span> <span class="p">},</span>
</span></span><span class="line"><span class="cl">        <span class="p">{</span> <span class="n">error</span> <span class="o">-&gt;</span> <span class="n">println</span><span class="p">(</span><span class="s2">&#34;error: </span><span class="si">$error</span><span class="s2">&#34;</span><span class="p">)</span> <span class="p">}</span>
</span></span><span class="line"><span class="cl">    <span class="p">)</span>
</span></span></code></pre></div><h2 id="kotlin-flow-error-handling">Kotlin Flow Error Handling</h2>
<p>As Kotlin Flow is essentially based on coroutines, the following applies:</p>
<ul>
<li>
<p>“Error” is a general Exception which can be thrown like any other exception</p>
</li>
<li>
<p>That “error” is propagated via general coroutines error handling mechanism (propagating to parent jobs and canceling all the jobs in the coroutine scope)</p>
</li>
<li>
<p>Consumers need to wrap Flow with try-catch if they want to handle exceptions</p>
</li>
<li>
<p>If the consumer didn’t handle errors with try-catch then that exception will be thrown as usual and be handled by the parent coroutine scope (it might immediately crash your app or be handled by exception handler e.g. in launch)</p>
</li>
<li>
<p>As we have to wrap chain in try-catch it looks like exception escaped the chain. It might be partially true, though as we’ll see later this is not the only mechanism to work with exceptions.</p>
</li>
</ul>
<h2 id="comparison">Comparison</h2>
<h3 id="setup">Setup</h3>
<p>First of all, for our comparison, we’ll create helper functions of Flow and Observable, which emit a value and then throw an exception.</p>
<p>For Observable:</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-kotlin" data-lang="kotlin"><span class="line"><span class="cl"><span class="k">private</span> <span class="k">fun</span> <span class="nf">observable</span><span class="p">(</span>
</span></span><span class="line"><span class="cl">    <span class="k">value</span><span class="p">:</span> <span class="n">Int</span> <span class="p">=</span> <span class="m">1</span>
</span></span><span class="line"><span class="cl"><span class="p">):</span> <span class="n">Observable</span><span class="p">&lt;</span><span class="n">Int</span><span class="p">&gt;</span> <span class="p">{</span>
</span></span><span class="line"><span class="cl">    <span class="k">return</span> <span class="nc">Observable</span><span class="p">.</span><span class="n">create</span> <span class="p">{</span> <span class="n">emitter</span> <span class="o">-&gt;</span>
</span></span><span class="line"><span class="cl">        <span class="n">emitter</span><span class="p">.</span><span class="n">onNext</span><span class="p">(</span><span class="k">value</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">        <span class="n">emitter</span><span class="p">.</span><span class="n">onError</span><span class="p">(</span><span class="n">RuntimeException</span><span class="p">())</span>
</span></span><span class="line"><span class="cl">    <span class="p">}</span>
</span></span><span class="line"><span class="cl"><span class="p">}</span>
</span></span></code></pre></div><p>And for Kotlin Flow (due to clash of names have to name it with “my” prefix):</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-kotlin" data-lang="kotlin"><span class="line"><span class="cl"><span class="k">private</span> <span class="k">fun</span> <span class="nf">myFlow</span><span class="p">(</span>
</span></span><span class="line"><span class="cl">    <span class="k">value</span><span class="p">:</span> <span class="n">Int</span> <span class="p">=</span> <span class="m">1</span>
</span></span><span class="line"><span class="cl"><span class="p">):</span> <span class="n">Flow</span><span class="p">&lt;</span><span class="n">Int</span><span class="p">&gt;</span> <span class="p">{</span>
</span></span><span class="line"><span class="cl">    <span class="k">return</span> <span class="n">flow</span> <span class="p">{</span>
</span></span><span class="line"><span class="cl">        <span class="n">emit</span><span class="p">(</span><span class="k">value</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">        <span class="k">throw</span> <span class="n">RuntimeException</span><span class="p">()</span>
</span></span><span class="line"><span class="cl">    <span class="p">}</span>
</span></span><span class="line"><span class="cl"><span class="p">}</span>
</span></span></code></pre></div><p>Next, we’ll set up short tests, which will be parametrized by operators. In that test we’ll take our test streams, subscribe to them on the io thread pool, then optionally will apply some operator and then print values and errors which will happen in the chain.</p>
<p>For Observable:</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-kotlin" data-lang="kotlin"><span class="line"><span class="cl"><span class="k">private</span> <span class="k">fun</span> <span class="nf">testObservable</span><span class="p">(</span>
</span></span><span class="line"><span class="cl">    <span class="k">operator</span><span class="p">:</span> <span class="n">Observable</span><span class="p">&lt;</span><span class="n">Int</span><span class="p">&gt;.()</span> <span class="o">-&gt;</span> <span class="n">Observable</span><span class="p">&lt;</span><span class="n">Int</span><span class="p">&gt;</span>
</span></span><span class="line"><span class="cl"><span class="p">)</span> <span class="p">{</span>
</span></span><span class="line"><span class="cl">    <span class="k">val</span> <span class="py">latch</span> <span class="p">=</span> <span class="n">CountDownLatch</span><span class="p">(</span><span class="m">1</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">    <span class="k">val</span> <span class="py">result</span> <span class="p">=</span> <span class="n">StringBuffer</span><span class="p">()</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="n">observable</span><span class="p">()</span>
</span></span><span class="line"><span class="cl">        <span class="p">.</span><span class="n">subscribeOn</span><span class="p">(</span><span class="nc">Schedulers</span><span class="p">.</span><span class="n">io</span><span class="p">())</span>
</span></span><span class="line"><span class="cl">        <span class="p">.</span><span class="k">operator</span><span class="p">()</span>
</span></span><span class="line"><span class="cl">        <span class="p">.</span><span class="n">doOnTerminate</span> <span class="p">{</span> <span class="n">latch</span><span class="p">.</span><span class="n">countDown</span><span class="p">()</span> <span class="p">}</span>
</span></span><span class="line"><span class="cl">        <span class="p">.</span><span class="n">subscribe</span><span class="p">(</span>
</span></span><span class="line"><span class="cl">            <span class="p">{</span> <span class="n">result</span><span class="p">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&#34;next </span><span class="si">$it</span><span class="s2">, &#34;</span><span class="p">)</span> <span class="p">},</span>
</span></span><span class="line"><span class="cl">            <span class="p">{</span> <span class="n">result</span><span class="p">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&#34;error </span><span class="si">$it</span><span class="s2">&#34;</span><span class="p">)</span> <span class="p">}</span>
</span></span><span class="line"><span class="cl">        <span class="p">)</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="n">latch</span><span class="p">.</span><span class="n">await</span><span class="p">()</span>
</span></span><span class="line"><span class="cl">    <span class="n">println</span><span class="p">(</span><span class="n">result</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="p">}</span>
</span></span></code></pre></div><p>For Kotlin Flow (note, that here we wrapped all the flow into the try-catch):</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-kotlin" data-lang="kotlin"><span class="line"><span class="cl"><span class="k">private</span> <span class="k">fun</span> <span class="nf">testFlow</span><span class="p">(</span>
</span></span><span class="line"><span class="cl">    <span class="k">operator</span><span class="p">:</span> <span class="n">Flow</span><span class="p">&lt;</span><span class="n">Int</span><span class="p">&gt;.()</span> <span class="o">-&gt;</span> <span class="n">Flow</span><span class="p">&lt;</span><span class="n">Int</span><span class="p">&gt;</span>
</span></span><span class="line"><span class="cl"><span class="p">)</span> <span class="p">{</span>
</span></span><span class="line"><span class="cl">    <span class="k">val</span> <span class="py">latch</span> <span class="p">=</span> <span class="n">CountDownLatch</span><span class="p">(</span><span class="m">1</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">    <span class="k">val</span> <span class="py">result</span> <span class="p">=</span> <span class="n">StringBuffer</span><span class="p">()</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl">    <span class="n">CoroutineScope</span><span class="p">(</span><span class="n">Job</span><span class="p">()</span> <span class="p">+</span> <span class="nc">Dispatchers</span><span class="p">.</span><span class="n">IO</span><span class="p">).</span><span class="n">launch</span> <span class="p">{</span>
</span></span><span class="line"><span class="cl">        <span class="k">try</span> <span class="p">{</span>
</span></span><span class="line"><span class="cl">            <span class="n">myFlow</span><span class="p">()</span>
</span></span><span class="line"><span class="cl">                <span class="p">.</span><span class="k">operator</span><span class="p">()</span>
</span></span><span class="line"><span class="cl">                <span class="p">.</span><span class="n">onCompletion</span> <span class="p">{</span> <span class="n">latch</span><span class="p">.</span><span class="n">countDown</span><span class="p">()</span> <span class="p">}</span>
</span></span><span class="line"><span class="cl">                <span class="p">.</span><span class="n">collect</span> <span class="p">{</span>
</span></span><span class="line"><span class="cl">                    <span class="n">result</span><span class="p">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&#34;next </span><span class="si">$it</span><span class="s2">, &#34;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">                <span class="p">}</span>
</span></span><span class="line"><span class="cl">        <span class="p">}</span> <span class="k">catch</span> <span class="p">(</span><span class="n">e</span><span class="p">:</span> <span class="n">Exception</span><span class="p">)</span> <span class="p">{</span>
</span></span><span class="line"><span class="cl">            <span class="n">result</span><span class="p">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&#34;error </span><span class="si">$e</span><span class="s2">&#34;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">        <span class="p">}</span>
</span></span><span class="line"><span class="cl">    <span class="p">}</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl">    <span class="n">latch</span><span class="p">.</span><span class="n">await</span><span class="p">()</span>
</span></span><span class="line"><span class="cl">    <span class="n">println</span><span class="p">(</span><span class="n">result</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="p">}</span>
</span></span></code></pre></div><h3 id="basic-error-handling">Basic Error Handling</h3>
<p>First, our test will be to see what is the default behavior without additional changes. In this case, we’ll provide an identity operator:</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-kotlin" data-lang="kotlin"><span class="line"><span class="cl"><span class="n">testObservable</span> <span class="p">{</span> <span class="k">this</span> <span class="p">}</span>
</span></span><span class="line"><span class="cl"><span class="n">testFlow</span> <span class="p">{</span> <span class="k">this</span> <span class="p">}</span>
</span></span></code></pre></div><p>In both cases in logs it will be printed:</p>
<pre tabindex="0"><code>next 1, error java.lang.RuntimeException
</code></pre><p>So, as we already discussed, our test streams emit one value and then terminate with an exception.</p>
<h3 id="catching-errors-with-an-emitting-default-value">Catching Errors with an emitting default value</h3>
<p>What if we don’t want our stream to be terminated (as error event terminates the stream)? One of the options is to emit some default value if we encounter an exception.</p>
<p>In RxJava for that, there is a special operator called onErrorReturn:</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-kotlin" data-lang="kotlin"><span class="line"><span class="cl"><span class="n">testObservable</span> <span class="p">{</span> <span class="n">onErrorReturn</span> <span class="p">{</span> <span class="m">5</span> <span class="p">}</span> <span class="p">}</span>
</span></span></code></pre></div><p>If we run this example we’ll see that in logs it will be printed:</p>
<pre tabindex="0"><code>next 1, next 5,
</code></pre><p>So the first item in the stream (1) was emitted as before, but then instead of an error, we have a default value (5) emitted as well.</p>
<p>In Kotlin Flow for the same use case, there is an operator catch, and the usage is the following:</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-kotlin" data-lang="kotlin"><span class="line"><span class="cl"><span class="n">testFlow</span> <span class="p">{</span> <span class="k">catch</span> <span class="p">{</span> <span class="n">emit</span><span class="p">(</span><span class="m">5</span><span class="p">)</span> <span class="p">}</span> <span class="p">}</span>
</span></span></code></pre></div><p>Operator catch catches all the exceptions from the upstream and can do some work, such as emitting default value. And result in logs will be the same.</p>
<h3 id="catching-errors-by-switching-to-another-stream">Catching Errors by switching to another stream</h3>
<p>Another option with not failing our stream on error is to switch to another stream. In RxJava for that, there is a special operator onErrorResumeNext:</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-kotlin" data-lang="kotlin"><span class="line"><span class="cl"><span class="n">testObservable</span> <span class="p">{</span> <span class="n">onErrorResumeNext</span><span class="p">(</span><span class="nc">Observable</span><span class="p">.</span><span class="n">just</span><span class="p">(</span><span class="m">1</span><span class="p">,</span> <span class="m">2</span><span class="p">,</span> <span class="m">3</span><span class="p">))</span> <span class="p">}</span>
</span></span></code></pre></div><p>After running the code the result will be:</p>
<pre tabindex="0"><code>next 1, next 1, next 2, next 3,
</code></pre><p>First, we have the value of 1 emitted (this is from the initial stream), then after error happened we switch to the new stream and all its values are emitted (and we won’t have an exception thrown).</p>
<blockquote>
<p><strong>NOTE</strong>: Essentially it is possible to achieve onErrorReturn with onErrorResumeNext by using onErrorResumeNext(Observable.just(5)), though usage of onErrorResumeNext has bigger overhead, so use it only when new stream is really needed.</p>
</blockquote>
<p>In Kotlin Flow for such a case we use that same operator catch:</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-kotlin" data-lang="kotlin"><span class="line"><span class="cl"><span class="n">testFlow</span> <span class="p">{</span> <span class="k">catch</span> <span class="p">{</span> <span class="n">emitAll</span><span class="p">(</span><span class="n">flowOf</span><span class="p">(</span><span class="m">1</span><span class="p">,</span> <span class="m">2</span><span class="p">,</span> <span class="m">3</span><span class="p">))</span> <span class="p">}</span> <span class="p">}</span>
</span></span></code></pre></div><p>Instead of emitting a single value on error, we emitAll values from the stream with the same expected result.</p>
<h3 id="intercepting-errors">Intercepting Errors</h3>
<p>The next thing which we’ll do is intercepting errors. When we intercept errors we can do some side effects without actually handling that error, for example, logging.</p>
<p>In RxJava for this case, we can use the doOnError operator:</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-kotlin" data-lang="kotlin"><span class="line"><span class="cl"><span class="n">testObservable</span> <span class="p">{</span> <span class="n">doOnError</span> <span class="p">{</span> <span class="n">print</span><span class="p">(</span><span class="s2">&#34;INTERCEPTED </span><span class="si">$it</span><span class="s2">, &#34;</span><span class="p">)</span> <span class="p">}</span> <span class="p">}</span>
</span></span></code></pre></div><p>The result will be:</p>
<pre tabindex="0"><code>INTERCEPTED java.lang.RuntimeException, next 1, error java.lang.RuntimeException
</code></pre><p>So, first, we intercepted our exception, printed some logs, and then as in the case with basic error handling, we got the first item emitted with error followed.</p>
<p>In Kotlin Flow though there is no special operator for that case, but we can write our own:</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-kotlin" data-lang="kotlin"><span class="line"><span class="cl"><span class="k">fun</span> <span class="p">&lt;</span><span class="nc">T</span><span class="p">&gt;</span> <span class="nf">Flow</span><span class="p">&lt;</span><span class="n">T</span><span class="p">&gt;.</span><span class="n">doOnError</span><span class="p">(</span><span class="n">onError</span><span class="p">:</span> <span class="p">(</span><span class="n">Throwable</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Unit</span><span class="p">):</span> <span class="n">Flow</span><span class="p">&lt;</span><span class="n">T</span><span class="p">&gt;</span> <span class="p">{</span>
</span></span><span class="line"><span class="cl">    <span class="k">return</span> <span class="n">flow</span> <span class="p">{</span>
</span></span><span class="line"><span class="cl">        <span class="k">try</span> <span class="p">{</span>
</span></span><span class="line"><span class="cl">            <span class="n">collect</span> <span class="p">{</span> <span class="k">value</span> <span class="o">-&gt;</span>
</span></span><span class="line"><span class="cl">                <span class="n">emit</span><span class="p">(</span><span class="k">value</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">            <span class="p">}</span>
</span></span><span class="line"><span class="cl">        <span class="p">}</span> <span class="k">catch</span> <span class="p">(</span><span class="n">e</span><span class="p">:</span> <span class="n">Exception</span><span class="p">)</span> <span class="p">{</span>
</span></span><span class="line"><span class="cl">            <span class="n">onError</span><span class="p">(</span><span class="n">e</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">            <span class="k">throw</span> <span class="n">e</span>
</span></span><span class="line"><span class="cl">        <span class="p">}</span>
</span></span><span class="line"><span class="cl">    <span class="p">}</span>
</span></span><span class="line"><span class="cl"><span class="p">}</span>
</span></span></code></pre></div><p>Here we just create new Flow, inside it, we start collecting all the values. We wrap everything in try-catch and if an error happens — we’ll execute the callback and re-throw exception.</p>
<p>The usage will be the following:</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-kotlin" data-lang="kotlin"><span class="line"><span class="cl"><span class="n">testFlow</span> <span class="p">{</span> <span class="n">doOnError</span> <span class="p">{</span> <span class="n">print</span><span class="p">(</span><span class="s2">&#34;INTERCEPTED </span><span class="si">$it</span><span class="s2">, &#34;</span><span class="p">)</span> <span class="p">}</span> <span class="p">}</span>
</span></span></code></pre></div><p>And as before, results in logs will be the same.</p>
<p>That’s all with the handling errors in the chain, but there is another topic of handling errors in inner streams.</p>
<h3 id="inner-streams-errors-handling">Inner streams errors handling</h3>
<p>As we know we can have inner streams by using flatMap in RxJava or flatMapMerge in Flow (or other xMap operators in RxJava or flatMapX operators in Kotlin Flow).
We can apply directly what we’ve learned to our inner streams, though sometimes there is a need to control error handling from the parent stream perspective.</p>
<p>We’ll modify our test samples a bit to better show that.</p>
<p>For Observable our new example will be:</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-kotlin" data-lang="kotlin"><span class="line"><span class="cl"><span class="k">private</span> <span class="k">fun</span> <span class="nf">testInnerObservable</span><span class="p">(</span>
</span></span><span class="line"><span class="cl">    <span class="k">operator</span><span class="p">:</span> <span class="n">Observable</span><span class="p">&lt;</span><span class="n">Int</span><span class="p">&gt;.()</span> <span class="o">-&gt;</span> <span class="n">Observable</span><span class="p">&lt;</span><span class="n">Int</span><span class="p">&gt;</span>
</span></span><span class="line"><span class="cl"><span class="p">)</span> <span class="p">{</span>
</span></span><span class="line"><span class="cl">    <span class="k">val</span> <span class="py">latch</span> <span class="p">=</span> <span class="n">CountDownLatch</span><span class="p">(</span><span class="m">1</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">    <span class="k">val</span> <span class="py">result</span> <span class="p">=</span> <span class="n">StringBuffer</span><span class="p">()</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl">    <span class="nc">Observable</span><span class="p">.</span><span class="n">just</span><span class="p">(</span><span class="m">10</span><span class="p">,</span> <span class="m">11</span><span class="p">,</span> <span class="m">12</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">        <span class="p">.</span><span class="n">subscribeOn</span><span class="p">(</span><span class="nc">Schedulers</span><span class="p">.</span><span class="n">io</span><span class="p">())</span>
</span></span><span class="line"><span class="cl">        <span class="p">.</span><span class="k">operator</span><span class="p">()</span>
</span></span><span class="line"><span class="cl">        <span class="p">.</span><span class="n">doOnTerminate</span> <span class="p">{</span> <span class="n">latch</span><span class="p">.</span><span class="n">countDown</span><span class="p">()</span> <span class="p">}</span>
</span></span><span class="line"><span class="cl">        <span class="p">.</span><span class="n">subscribe</span><span class="p">(</span>
</span></span><span class="line"><span class="cl">            <span class="p">{</span> <span class="n">result</span><span class="p">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&#34;next </span><span class="si">$it</span><span class="s2">, &#34;</span><span class="p">)</span> <span class="p">},</span>
</span></span><span class="line"><span class="cl">            <span class="p">{</span> <span class="n">result</span><span class="p">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&#34;error </span><span class="si">$it</span><span class="s2">&#34;</span><span class="p">)</span> <span class="p">}</span>
</span></span><span class="line"><span class="cl">        <span class="p">)</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl">    <span class="n">latch</span><span class="p">.</span><span class="n">await</span><span class="p">()</span>
</span></span><span class="line"><span class="cl">    <span class="n">println</span><span class="p">(</span><span class="n">result</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="p">}</span>
</span></span></code></pre></div><p>The difference is that now our parent observable won’t throw an exception, instead it will emit three items and complete. Inside operator we’ll be able to add our inner streams.</p>
<p>For Kotlin Flow we’ll have the following sample:</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-kotlin" data-lang="kotlin"><span class="line"><span class="cl"><span class="k">private</span> <span class="k">fun</span> <span class="nf">testInnerFlow</span><span class="p">(</span>
</span></span><span class="line"><span class="cl">    <span class="k">operator</span><span class="p">:</span> <span class="n">Flow</span><span class="p">&lt;</span><span class="n">Int</span><span class="p">&gt;.()</span> <span class="o">-&gt;</span> <span class="n">Flow</span><span class="p">&lt;</span><span class="n">Int</span><span class="p">&gt;</span>
</span></span><span class="line"><span class="cl"><span class="p">)</span> <span class="p">{</span>
</span></span><span class="line"><span class="cl">    <span class="k">val</span> <span class="py">latch</span> <span class="p">=</span> <span class="n">CountDownLatch</span><span class="p">(</span><span class="m">1</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">    <span class="k">val</span> <span class="py">result</span> <span class="p">=</span> <span class="n">StringBuffer</span><span class="p">()</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl">    <span class="n">CoroutineScope</span><span class="p">(</span><span class="n">Job</span><span class="p">()</span> <span class="p">+</span> <span class="nc">Dispatchers</span><span class="p">.</span><span class="n">IO</span><span class="p">).</span><span class="n">launch</span> <span class="p">{</span>
</span></span><span class="line"><span class="cl">        <span class="k">try</span> <span class="p">{</span>
</span></span><span class="line"><span class="cl">            <span class="n">flowOf</span><span class="p">(</span><span class="m">10</span><span class="p">,</span> <span class="m">11</span><span class="p">,</span> <span class="m">12</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">                <span class="p">.</span><span class="k">operator</span><span class="p">()</span>
</span></span><span class="line"><span class="cl">                <span class="p">.</span><span class="n">onCompletion</span> <span class="p">{</span> <span class="n">latch</span><span class="p">.</span><span class="n">countDown</span><span class="p">()</span> <span class="p">}</span>
</span></span><span class="line"><span class="cl">                <span class="p">.</span><span class="n">collect</span> <span class="p">{</span>
</span></span><span class="line"><span class="cl">                    <span class="n">result</span><span class="p">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&#34;next </span><span class="si">$it</span><span class="s2">, &#34;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">                <span class="p">}</span>
</span></span><span class="line"><span class="cl">        <span class="p">}</span> <span class="k">catch</span> <span class="p">(</span><span class="n">e</span><span class="p">:</span> <span class="n">Exception</span><span class="p">)</span> <span class="p">{</span>
</span></span><span class="line"><span class="cl">            <span class="n">result</span><span class="p">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&#34;error </span><span class="si">$e</span><span class="s2">&#34;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">        <span class="p">}</span>
</span></span><span class="line"><span class="cl">    <span class="p">}</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl">    <span class="n">latch</span><span class="p">.</span><span class="n">await</span><span class="p">()</span>
</span></span><span class="line"><span class="cl">    <span class="n">println</span><span class="p">(</span><span class="n">result</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="p">}</span>
</span></span></code></pre></div><p>There are many different operators in RxJava, but we’ll take a look only at one: concatMapDelayError.
In order to understand how it works, let’s first look at the concatMap and flatMapConcat operators. Both these operators start subscribing to the inner streams one by one. And subscription to the next inner stream happens only when previous was completed.</p>
<p>For RxJava and Kotlin Flow it will look like:</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-kotlin" data-lang="kotlin"><span class="line"><span class="cl"><span class="n">testInnerObservable</span> <span class="p">{</span> <span class="n">concatMap</span> <span class="p">{</span> <span class="n">observable</span><span class="p">(</span><span class="k">it</span><span class="p">)</span> <span class="p">}</span> <span class="p">}</span>
</span></span><span class="line"><span class="cl"><span class="n">testInnerFlow</span> <span class="p">{</span> <span class="n">flatMapConcat</span> <span class="p">{</span> <span class="n">myFlow</span><span class="p">(</span><span class="k">it</span><span class="p">)</span> <span class="p">}</span> <span class="p">}</span>
</span></span></code></pre></div><p>What happens here:</p>
<ul>
<li>
<p>Our parent stream has three values</p>
</li>
<li>
<p>For each value emitted we switch to the new inner stream (which emits value and then fails)</p>
</li>
<li>
<p>Subscription to the next inner stream happens only after the previous inner stream has completed</p>
</li>
</ul>
<p>Result of executing for RxJava and Kotlin Flow will be the same:</p>
<pre tabindex="0"><code>next 10, error java.lang.RuntimeException
</code></pre><p>We have first item emitted followed by an error and no more values are emitted. This happens because the first item (10) triggered a subscription of the inner stream, which emitted that value (10) and then failed with an exception. Because of that, the whole stream failed and the next item from parent (11) hasn’t been switched to the new inner stream.</p>
<p>What if we want to have all the values emitted and emit error only in the end (if it happened)? For such a case there is concatMapDelayError in RxJava. It will delay all the errors until the stream completes and throw CompositeException with all the exceptions which happened in the stream.</p>
<p>If we run the following code:</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-kotlin" data-lang="kotlin"><span class="line"><span class="cl"><span class="n">testInnerObservable</span> <span class="p">{</span> <span class="n">concatMapDelayError</span> <span class="p">{</span> <span class="n">observable</span><span class="p">(</span><span class="k">it</span><span class="p">)</span> <span class="p">}</span> <span class="p">}</span>
</span></span></code></pre></div><p>We see the result:</p>
<pre tabindex="0"><code>next 10, next 11, next 12, error io.reactivex.exceptions.CompositeException: 3 exceptions occurred.
</code></pre><p>All three items were emitted followed by a single composite exception (which has three exceptions inside — one per each inner stream).</p>
<p>What about Kotlin Flow? It doesn’t have any operator with delaying errors. But we can try to write our own. We’ll start by copy-pasting sources of flatMapConcat operator and add some error handling there:</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-kotlin" data-lang="kotlin"><span class="line"><span class="cl"><span class="k">fun</span> <span class="p">&lt;</span><span class="nc">T</span><span class="p">,</span> <span class="nc">R</span><span class="p">&gt;</span> <span class="nf">Flow</span><span class="p">&lt;</span><span class="n">T</span><span class="p">&gt;.</span><span class="n">flatMapConcatDelayError</span><span class="p">(</span>
</span></span><span class="line"><span class="cl">    <span class="n">transform</span><span class="p">:</span> <span class="k">suspend</span> <span class="p">(</span><span class="k">value</span><span class="p">:</span> <span class="n">T</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Flow</span><span class="p">&lt;</span><span class="n">R</span><span class="p">&gt;</span>
</span></span><span class="line"><span class="cl"><span class="p">):</span> <span class="n">Flow</span><span class="p">&lt;</span><span class="n">R</span><span class="p">&gt;</span> <span class="p">=</span> <span class="n">map</span><span class="p">(</span><span class="n">transform</span><span class="p">).</span><span class="n">flattenConcatDelayError</span><span class="p">()</span>
</span></span></code></pre></div><p>We’ll start from defining our operator flatMapConcatDelayError as combination of map and flatten (yep, map + flatten == flatMap).
Then we’ll define our flattenConcatDelayError:</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-kotlin" data-lang="kotlin"><span class="line"><span class="cl"><span class="k">fun</span> <span class="p">&lt;</span><span class="nc">T</span><span class="p">&gt;</span> <span class="nf">Flow</span><span class="p">&lt;</span><span class="n">Flow</span><span class="p">&lt;</span><span class="n">T</span><span class="p">&gt;&gt;.</span><span class="n">flattenConcatDelayError</span><span class="p">():</span> <span class="n">Flow</span><span class="p">&lt;</span><span class="n">T</span><span class="p">&gt;</span> <span class="p">{</span>
</span></span><span class="line"><span class="cl">    <span class="k">val</span> <span class="py">list</span> <span class="p">=</span> <span class="n">CopyOnWriteArrayList</span><span class="p">&lt;</span><span class="n">Exception</span><span class="p">&gt;()</span>
</span></span><span class="line"><span class="cl">    <span class="k">return</span> <span class="n">flow</span><span class="p">&lt;</span><span class="n">T</span><span class="p">&gt;</span> <span class="p">{</span>
</span></span><span class="line"><span class="cl">        <span class="n">collect</span> <span class="p">{</span> <span class="k">value</span> <span class="o">-&gt;</span>
</span></span><span class="line"><span class="cl">            <span class="k">try</span> <span class="p">{</span>
</span></span><span class="line"><span class="cl">                <span class="n">emitAll</span><span class="p">(</span><span class="k">value</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">            <span class="p">}</span> <span class="k">catch</span> <span class="p">(</span><span class="n">e</span><span class="p">:</span> <span class="n">Exception</span><span class="p">)</span> <span class="p">{</span>
</span></span><span class="line"><span class="cl">                <span class="n">list</span><span class="p">.</span><span class="n">add</span><span class="p">(</span><span class="n">e</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">            <span class="p">}</span>
</span></span><span class="line"><span class="cl">        <span class="p">}</span>
</span></span><span class="line"><span class="cl">    <span class="p">}.</span><span class="n">onCompletion</span> <span class="p">{</span>
</span></span><span class="line"><span class="cl">        <span class="k">if</span> <span class="p">(</span><span class="n">list</span><span class="p">.</span><span class="n">isNotEmpty</span><span class="p">())</span> <span class="p">{</span>
</span></span><span class="line"><span class="cl">            <span class="k">throw</span> <span class="n">CompositeException</span><span class="p">(</span><span class="n">list</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">        <span class="p">}</span>
</span></span><span class="line"><span class="cl">    <span class="p">}</span>
</span></span><span class="line"><span class="cl"><span class="p">}</span>
</span></span></code></pre></div><p>What we do here is wrap our emits into try-catch with saving all the exceptions (note that we need to use thread-safe list). And throw a composite exception at the end if there were some exceptions.</p>
<p>Final call for Kotlin Flow will be:</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-kotlin" data-lang="kotlin"><span class="line"><span class="cl"><span class="n">testInnerFlow</span> <span class="p">{</span> <span class="n">flatMapConcatDelayError</span> <span class="p">{</span> <span class="n">myFlow</span><span class="p">(</span><span class="k">it</span><span class="p">)</span> <span class="p">}</span> <span class="p">}</span>
</span></span></code></pre></div><h3 id="rxjava-like-subscribe-for-kotlin-flow">RxJava-like subscribe for Kotlin Flow</h3>
<p>Do I need to always use either catch at the bottom of each chain or wrap everything into try-catch? I’d say yes, but probably it could be useful to have similar API as in RxJava and have single subscribe with three lambdas.</p>
<p>We can try to implement such:</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-kotlin" data-lang="kotlin"><span class="line"><span class="cl"><span class="k">suspend</span> <span class="k">fun</span> <span class="p">&lt;</span><span class="nc">T</span><span class="p">&gt;</span> <span class="nf">Flow</span><span class="p">&lt;</span><span class="n">T</span><span class="p">&gt;.</span><span class="n">subscribe</span><span class="p">(</span>
</span></span><span class="line"><span class="cl">    <span class="n">onNext</span><span class="p">:</span> <span class="p">(</span><span class="n">T</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Unit</span><span class="p">,</span>
</span></span><span class="line"><span class="cl">    <span class="n">onError</span><span class="p">:</span> <span class="p">(</span><span class="n">Throwable</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Unit</span><span class="p">,</span>
</span></span><span class="line"><span class="cl">    <span class="n">onComplete</span><span class="p">:</span> <span class="p">()</span> <span class="o">-&gt;</span> <span class="n">Unit</span>
</span></span><span class="line"><span class="cl"><span class="p">)</span> <span class="p">{</span>
</span></span><span class="line"><span class="cl">    <span class="k">this</span>
</span></span><span class="line"><span class="cl">        <span class="p">.</span><span class="n">onEach</span> <span class="p">{</span> <span class="n">onNext</span><span class="p">(</span><span class="k">it</span><span class="p">)</span> <span class="p">}</span>
</span></span><span class="line"><span class="cl">        <span class="p">.</span><span class="n">onCompletion</span> <span class="p">{</span> <span class="n">error</span><span class="p">:</span> <span class="n">Throwable</span><span class="p">?</span> <span class="o">-&gt;</span>
</span></span><span class="line"><span class="cl">            <span class="k">if</span> <span class="p">(</span><span class="n">error</span> <span class="o">==</span> <span class="k">null</span><span class="p">)</span> <span class="p">{</span>
</span></span><span class="line"><span class="cl">                <span class="n">onComplete</span><span class="p">()</span>
</span></span><span class="line"><span class="cl">            <span class="p">}</span>
</span></span><span class="line"><span class="cl">        <span class="p">}</span>
</span></span><span class="line"><span class="cl">        <span class="p">.</span><span class="k">catch</span> <span class="p">{</span> <span class="n">onError</span><span class="p">(</span><span class="k">it</span><span class="p">)</span> <span class="p">}</span>
</span></span><span class="line"><span class="cl">        <span class="p">.</span><span class="n">collect</span><span class="p">()</span>
</span></span><span class="line"><span class="cl"><span class="p">}</span>
</span></span></code></pre></div><p>What we do here is instead of terminal collect method, we’ll use our own subscribe, which takes three lambdas.
In onEach we’ll get all the values emitted. In onCompletion we’ll see that we track our onComplete (note that we need to check that we completed without error). And in catch we catch all the exceptions.</p>
<blockquote>
<p><strong>NOTE</strong>: we could move onError inside onCompletion if we wish and make catch clause empty. Say it is matter of taste.</p>
</blockquote>
<p><strong>But did anyone spot the issue?</strong></p>
<p>Let’s run few tests:</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-kotlin" data-lang="kotlin"><span class="line"><span class="cl"><span class="n">runBlocking</span> <span class="p">{</span>
</span></span><span class="line"><span class="cl">    <span class="n">flowOf</span><span class="p">(</span><span class="m">1</span><span class="p">,</span> <span class="m">2</span><span class="p">,</span> <span class="m">3</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">        <span class="p">.</span><span class="n">subscribe</span><span class="p">(</span>
</span></span><span class="line"><span class="cl">            <span class="p">{</span> <span class="n">print</span><span class="p">(</span><span class="s2">&#34;next </span><span class="si">$it</span><span class="s2">, &#34;</span><span class="p">)</span> <span class="p">},</span>
</span></span><span class="line"><span class="cl">            <span class="p">{</span> <span class="n">print</span><span class="p">(</span><span class="s2">&#34;error </span><span class="si">$it</span><span class="s2">, &#34;</span><span class="p">)</span> <span class="p">},</span>
</span></span><span class="line"><span class="cl">            <span class="p">{</span> <span class="n">print</span><span class="p">(</span><span class="s2">&#34;complete &#34;</span><span class="p">)</span> <span class="p">}</span>
</span></span><span class="line"><span class="cl">        <span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="p">}</span>
</span></span></code></pre></div><p>It prints:</p>
<pre tabindex="0"><code>next 1, next 2, next 3, complete
</code></pre><p>So far so good. We have all three values emitted following by completion event.</p>
<p>Next test will be with error:</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-kotlin" data-lang="kotlin"><span class="line"><span class="cl"><span class="n">runBlocking</span> <span class="p">{</span>
</span></span><span class="line"><span class="cl">    <span class="n">flow</span> <span class="p">{</span>
</span></span><span class="line"><span class="cl">        <span class="n">emit</span><span class="p">(</span><span class="m">1</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">        <span class="k">throw</span> <span class="n">RuntimeException</span><span class="p">()</span>
</span></span><span class="line"><span class="cl">    <span class="p">}</span>
</span></span><span class="line"><span class="cl">        <span class="p">.</span><span class="n">subscribe</span><span class="p">(</span>
</span></span><span class="line"><span class="cl">            <span class="p">{</span> <span class="n">print</span><span class="p">(</span><span class="s2">&#34;next </span><span class="si">$it</span><span class="s2">, &#34;</span><span class="p">)</span> <span class="p">},</span>
</span></span><span class="line"><span class="cl">            <span class="p">{</span> <span class="n">print</span><span class="p">(</span><span class="s2">&#34;error </span><span class="si">$it</span><span class="s2">, &#34;</span><span class="p">)</span> <span class="p">},</span>
</span></span><span class="line"><span class="cl">            <span class="p">{</span> <span class="n">print</span><span class="p">(</span><span class="s2">&#34;complete &#34;</span><span class="p">)</span> <span class="p">}</span>
</span></span><span class="line"><span class="cl">        <span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="p">}</span>
</span></span></code></pre></div><p>It prints:</p>
<pre tabindex="0"><code>next 1, complete error java.lang.RuntimeException,
</code></pre><p>Also good, we have one value followed by error event. And we don’t have a completion event.</p>
<p>Let’s check the final test — we throw exception in a callback:</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-kotlin" data-lang="kotlin"><span class="line"><span class="cl"><span class="n">runBlocking</span> <span class="p">{</span>
</span></span><span class="line"><span class="cl">    <span class="n">flow</span> <span class="p">{</span>
</span></span><span class="line"><span class="cl">        <span class="n">emit</span><span class="p">(</span><span class="m">1</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">    <span class="p">}</span>
</span></span><span class="line"><span class="cl">        <span class="p">.</span><span class="n">subscribe</span><span class="p">(</span>
</span></span><span class="line"><span class="cl">            <span class="p">{</span> <span class="n">print</span><span class="p">(</span><span class="s2">&#34;next </span><span class="si">$it</span><span class="s2">, &#34;</span><span class="p">)</span> <span class="p">},</span>
</span></span><span class="line"><span class="cl">            <span class="p">{</span> <span class="n">print</span><span class="p">(</span><span class="s2">&#34;error </span><span class="si">$it</span><span class="s2">, &#34;</span><span class="p">)</span> <span class="p">},</span>
</span></span><span class="line"><span class="cl">            <span class="p">{</span> <span class="k">throw</span> <span class="n">RuntimeException</span><span class="p">()</span> <span class="p">}</span>
</span></span><span class="line"><span class="cl">        <span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="p">}</span>
</span></span></code></pre></div><p>The result will be:</p>
<pre tabindex="0"><code>next 1, error java.lang.RuntimeException,
</code></pre><p>And this is not what we’ve expected. If we have an exception in a callback, then it should be thrown and should not pass to our error handler. Error handler should catch all the errors in the chain and callback is not essentially part of that chain.</p>
<p>Why it happens so? Because catch is the last operator in the chain and it handles everything which is above. Including what is inside onEach and onCompletion operators.</p>
<p>This is the difference which is needed to keep in mind. If anyone would be able to create an identical implementation of subscribe method, feel free to post it in comments.</p>
<h2 id="upd-retry">UPD: retry</h2>
<p>As was pointed by feedback I missed in this article operators for retry logic. Basically both RxJava and Kotlin Flow has built-in support for retry. For example to retry stream twice (if there was error) one need to use “retry(2)” operator.</p>
<h2 id="conclusion">Conclusion</h2>
<p>Following this article, I hope we now have a better understanding of error handling in RxJava and Kotlin Flow. RxJava has a lot of operators, many of them are really useful (and are missing in Kotlin Flow right now).</p>
<p>We can write our own implementations, but they stick to be error-prone and platform-specific. Say that flatMapConcatDelayError operator: it uses platform-specific thread-safe list, which might work incorrectly with coroutines (which can jump over threads) — mean that mixing concepts might lead to weird issues. Also, the whole implementation might not be that refined, contain races or whatever.</p>
<p>It is possible to do everything we can do in RxJava with Kotlin Flow and hopefully, additional operators will be added to Kotlin Flow as well.</p>
<p>If anyone wants to play with the setup one can find in <a href="https://gist.github.com/krossovochkin/1a47f05d3ccbf20fffa872b22362e16d">this gist</a>. Feel free to post comments and share your examples or ideas on the important things in error handling in RxJava and Kotlin Flow.</p>
<p>Happy coding!</p>

</main>


<hr/>

<footer>
<nav>
  <ul>
    
    <li class="menu-left"><a href="/about">About</a></li>
    
    <li class="menu-left"><a href="/">Posts</a></li>
    
    <li class="menu-left"><a href="/apps">Apps</a></li>
    
    
    <li class="menu-right"><a rel="me" href="https://t.me/krossovochkin_newsletter" target="_blank"><i class="fa-brands fa-telegram fa-l"></i></a></li>
    
    <li class="menu-right"><a rel="me" href="https://stackoverflow.com/users/1533933/krossovochkin" target="_blank"><i class="fa-brands fa-stack-overflow fa-l"></i></a></li>
    
    <li class="menu-right"><a rel="me" href="https://linkedin.com/in/vasyadrobushkov" target="_blank"><i class="fa-brands fa-linkedin fa-l"></i></a></li>
    
    <li class="menu-right"><a rel="me" href="https://facebook.com/vasya.drobushkov" target="_blank"><i class="fa-brands fa-facebook fa-l"></i></a></li>
    
    <li class="menu-right"><a rel="me" href="/index.xml" target="_blank"><i class="fa-solid fa-rss fa-l"></i></a></li>
    
    <li class="menu-right"><a rel="me" href="https://androiddev.social/@krossovochkin" target="_blank"><i class="fa-brands fa-mastodon fa-l"></i></a></li>
    
    <li class="menu-right"><a rel="me" href="https://github.com/krossovochkin" target="_blank"><i class="fa-brands fa-github fa-l"></i></a></li>
    
  </ul>
</nav>

</footer>


</body>

</html>